cli/command/utils.go

```diff
@@ -16,3 +26,62 @@ func extractArgs(received []string, names ...string) (map[string]string, error)
 
 	return args, nil
 }
+
+func buildLogQueryHelper(id, start, end string, tail int) url.Values {
+	query := url.Values{}
+
+	if tail > 0 {
+		query.Set(client.LogQueryParamTail, strconv.Itoa(tail))
+	}
+
+	if start != "" {
+		query.Set(client.LogQueryParamStart, start)
+	}
+
+	if end != "" {
+		query.Set(client.LogQueryParamEnd, end)
+	}
+
+	return query
+}
+
+func SetTimeMultiplier(v time.Duration) func() {
+	timeMultiplier = v
+	return func() { timeMultiplier = 1 }
+}
+
+func WaitForDeployment(client client.Client, serviceID string, timeout time.Duration) (*models.Service, error) {
+	successCount := 0
+	requiredSuccessCount := 3
+
+	check := func() (bool, error) {
+		service, err := client.ReadService(serviceID)
+		if err != nil {
+			return false, err
+		}
+
+		for _, deployment := range service.Deployments {
+			if deployment.DesiredCount != deployment.RunningCount {
+				return false, nil
+			}
+		}
+
+		successCount++
+		return successCount >= requiredSuccessCount, nil
+	}
+
+	delay := 5 * time.Second * timeMultiplier
+
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(delay) {
```

Comment:
I like the sleeper pattern at https://github.com/quintilesims/layer0/blob/cli-refactor/client/helpers.go#L58. 
Perhaps we can incorporate that here? Also, I think the code can be improved a bit to avoid flapping since the current check only cares about 3 successes total, not 3 in a row:

```
var consecutiveSuccesses int
sleep := newLinearBackoffSleeper(time.Second)
for start := time.Now(); time.Since(start) < timeout; sleep() {
    service, err := client.ReadService(serviceID)
    ...

    for _, deployment := range service.Deployments {
        if deployment.DesiredCount != deployment.RunningCount {
            consecutiveSuccesses = 0
            continue
        }
    }

    consecutiveSuccesses++
    if consecutiveSuccesses >= 3 {
        return service, nil
    }
}

return nil, fmt.Errorf(...)
```

Reply:
✅

Reply:
@zpatrick I've taken those things into consideration, and refactored accordingly (see `layer0/client/helpers.go:WaitForDeployment()`). However, I kept my implementation using that `check()` function instead of this suggestion with `continue`, since that `continue` repeats the inner `for` loop over `service.Deployments` instead of the outer `time` loop. This has two undesired effects:
- `Deployment`s continue to be checked for validation after a failure has been detected.
- The `consecutiveSuccess++` operation happens for every iteration of the `time` loop, not just the iterations where the `*Count`s match.